qrenderdoc/Windows/ShaderViewer.cpp

```diff
@@ -838,6 +838,41 @@ void ShaderViewer::debugShader(const ShaderReflection *shader, ResourceId pipeli
       ui->execForwards->setMenu(forwardsMenu);
     }
 
+    {
+      QMenu *bookmarkMenu = new QMenu(this);
+      bookmarkMenu->setToolTipsVisible(true);
+      QAction *act;
+
+      act = MakeExecuteAction(tr("Toggle Bookmark"), Icons::bookmark_blue(),
+                              tr("Toggle bookmark on current line"),
+                              QKeySequence(Qt::Key_F2 | Qt::ControlModifier));
+
+      QObject::connect(act, &QAction::triggered, [this]() { ToggleBookmark(); });
+      bookmarkMenu->addAction(act);
+
+      act = MakeExecuteAction(tr("Next Bookmark"), Icons::bookmark_next(),
+                              tr("Go to next bookmark in the current file"),
+                              QKeySequence(Qt::Key_F2));
+
+      QObject::connect(act, &QAction::triggered, [this]() { NextBookmark(); });
+      bookmarkMenu->addAction(act);
+
+      act = MakeExecuteAction(tr("Previous Bookmark"), Icons::bookmark_previous(),
+                              tr("Go to previous bookmark in the current file"),
+                              QKeySequence(Qt::Key_F2 | Qt::ShiftModifier));
+
+      QObject::connect(act, &QAction::triggered, [this]() { PreviousBookmark(); });
+      bookmarkMenu->addAction(act);
+
+      act = MakeExecuteAction(tr("Clear All Bookmarks"), Icons::bookmark_remove(),
+                              tr("Clear all bookmarks in the current file"), QKeySequence());
+
+      QObject::connect(act, &QAction::triggered, [this]() { ClearAllBookmarks(); });
+      bookmarkMenu->addAction(act);
+
+      ui->bookmark->setMenu(bookmarkMenu);
+    }
```

Comment:
I think here we should mark the previous/next/clear buttons as disabled when there are no bookmarks in the current file as a typical UI convention.

You can do this with the `QMenu::aboutToShow` slot on the menu, if you keep the actions as separate named variables and capture them in a lambda then conditionally call `setEnabled`.

Reply:
Oh yeah I'll do that! I did it for the context menu but forgot about this one!